internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,367 @@
+	}
+
+	if err := w.catacomb.Add(connRequestWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case changes := <-connRequestWatcher.Changes():
+			for _, connId := range changes {
+				requestContext := w.catacomb.Context(context.Background())
+
+				err := w.handleConnection(requestContext, connId)
+				if err != nil {
+					w.logger.Errorf("Failed to handle connection %q: %v", connId, err)
+				}
+			}
+		}
+	}
+}
+
+// Kill implements the worker.Worker interface.
+func (w *sshSessionWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait implements the worker.Worker interface.
+func (w *sshSessionWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// handleConnection handles initiating a reverse SSH connection to the controller
+// and piping it to the local sshd of the machine agent's machine.
+//
+// This function does the following:
+//  1. Gets the controllers address and ephemeral public key for the connection.
+//  2. Verifies the address is known to this machine agent.
+//  3. Adds the ephemeral public key.
+//  4. Dials the controllers SSH server expecting an SSH connection to come
+//     back from the controller's SSH server.
+//  5. Pipes the connection to the local sshd.
+//  6. On connection close, removes the ephemeral public key.
+func (w *sshSessionWorker) handleConnection(ctx context.Context, connID string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		reqParams, err := w.facadeClient.GetSSHConnRequest(connID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		ctrlAddress := reqParams.ControllerAddresses.Values()[0]
+		ephemeralPublicKey := string(reqParams.EphemeralPublicKey)
+
+		if err := w.ephemeralKeysUpdater.AddEphemeralKey(ephemeralPublicKey); err != nil {
+			return errors.Trace(err)
+		}
+
+		defer func() {
+			if err := w.ephemeralKeysUpdater.RemoveEphemeralKey(ephemeralPublicKey); err != nil {
+				w.logger.Errorf("Error cleaning up ephemeral public key: %v", err)
+			}
+		}()
+
+		if err := w.pipeConnectionToSSHD(ctx, ctrlAddress, reqParams.Password); err != nil {
+			return errors.Trace(err)
+		}
+
+		return nil
+	}
+}
+
+// pipeConnectionToSSHD initiates the connection back to the controller and pipes
+// it over to the local SSHD. This call blocks until the connection has finished.
+func (w *sshSessionWorker) pipeConnectionToSSHD(ctx context.Context, ctrlAddress, password string) error {
+	controllerConn, err := w.connectionGetter.GetControllerConnection(password, ctrlAddress)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer controllerConn.Close()
+
+	sshdConn, err := w.connectionGetter.GetSSHDConnection()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer sshdConn.Close()
+
+	cancellableControllerConnection := newCancellableReadWriter(ctx, controllerConn)
+	cancellableSSHDConnection := newCancellableReadWriter(ctx, sshdConn)
+	eg := errgroup.Group{}
```

Comment:
Since this doesn't work as discussed, we can do something like
` go func(){
<-ctx.Done()
sshdConn.Close()
controllerConn.Close()
}()
```
But we need to make sure this routine is cleaned up when the IO copy finishes even if the ctx is not cancelled.